src/main/java/org/treadmill/zk/plugin/matcher/RoleMatcher.java

```diff
@@ -0,0 +1,49 @@
+package org.treadmill.zk.plugin.matcher;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import com.unboundid.ldap.sdk.LDAPSearchException;
+import org.slf4j.Logger;
+import org.treadmill.zk.plugin.utils.LdapQuery;
+
+import java.io.IOException;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+
+import static org.slf4j.LoggerFactory.getLogger;
+import static org.treadmill.zk.plugin.utils.Configuration.get;
+
+@Singleton
+public class RoleMatcher extends Matcher {
+
+  private static final Logger logger = getLogger(RoleMatcher.class);
+  private LoadingCache<String, Set<String>> cache;
+
+
+  @Inject
+  public RoleMatcher(LdapQuery ldapQuery) throws IOException {
+    setupCache(ldapQuery);
+  }
+
+  @Override
+  public boolean matchAcl(String principal, String role) throws ExecutionException {
+    if ("readers".equals(role)) return true;
```

Comment:
Since there isn't much for readers role and admin and servers role are quite similar, we went away from separate role matchers.
The admins entries in LDAP doesn't have "host/" prefix. We are simply caching whatever value is there in LDAP. Which means it will be without host for both hosts and treadmill user (treadmld).